src/extblkdev/ExtBlkDevPlugin.cc

```diff
@@ -0,0 +1,218 @@
+	      rc=-ENOENT;
+	      break;
+	    }
+	    rc=ebdplugin->get_required_cap_set(caps);
+	    if(rc!=0)
+	      break;
+	    // iterate over capabilities and check for active bits
+	    for(int i=0;i<=CAP_LAST_CAP;++i){
+	      cap_flag_value_t val;
+	      if(cap_get_flag(caps, i, CAP_PERMITTED, &val)<0){
+		rc=-errno;
+		break;
+	      }
+	      if(val!=CAP_CLEAR){
+		cap_value_t arr[1];
+		arr[0]=i;
+		// set capability in merged set
+		if(cap_set_flag(merge_caps, CAP_PERMITTED, 1, arr, CAP_SET)<0){
+		  rc=-errno;
+		  break;
+		}
+	      }
+	    }
+	    if(rc!=0)
+	      break;
+	  }
+	}
+	cap_free(caps);
+	if(rc!=0)
+	  break;
+
+	bool changed=false;
+	// get process capability set
+	cap_t proc_caps=cap_get_proc(); 
+	if(proc_caps==NULL){
+	  rc=-errno;
+	  break;
+	}
+	do{
+	  // iterate over capabilities
+	  for(int i=0;i<=CAP_LAST_CAP;++i){
+	    cap_flag_value_t val;
+	    if(cap_get_flag(merge_caps, i, CAP_PERMITTED, &val)<0){
+	      rc=-errno;
+	      break;
+	    }
+	    if(val==CAP_CLEAR){
+	      if(cap_get_flag(proc_caps, i, CAP_PERMITTED, &val)<0){
+		rc=-errno;
+		break;
+	      }
+	      if(val!=CAP_CLEAR){
+		// if bit clear in merged set, but set in process set, clear in process set
+		changed=true;
+		cap_value_t arr[1];
+		arr[0]=i;
+		if(cap_set_flag(proc_caps, CAP_PERMITTED, 1, arr, CAP_CLEAR)<0){
+		  rc=-errno;
+		  break;
+		}
+	      }
+	    }
+	  }
+	  // apply reduced capability set to process
+	  if(rc==0 && changed){
+	    if (cap_set_proc(proc_caps)<0){
+	      rc=-errno;
+	    }
+	  }
+	}while(0);
+	cap_free(proc_caps);
+      }while(0);
+      cap_free(merge_caps);
+#endif
+      return rc;
+    }
+
+    // preload set of extblkdev plugins defined in config
+    int preload(CephContext *cct)
+    {
+      const auto& conf = cct->_conf;
+      string plugins = conf.get_val<std::string>("osd_extblkdev_plugins");
+
+      list<string> plugins_list;
+      get_str_list(plugins, plugins_list);
+
+      auto registry=cct->get_plugin_registry();
+      std::lock_guard l(registry->lock);
+      for (auto& i : plugins_list) {
+	int rc = registry->load("extblkdev", std::string("ebd_")+i);
+	if (rc)
+	  return rc;
+      }
+      int rc=limit_caps(cct);
+      return rc;
+    }
+
+
+    // scan extblkdev plugins for support of this device
+    int detect_device(CephContext *cct,
+		      const std::string &logdevname,
+		      ExtBlkDevInterfaceRef& ebd_impl)
+    {
+      int rc=-ENOENT;
+      std::string plg_name;
+      auto registry=cct->get_plugin_registry();
+      auto ptype=registry->plugins.find("extblkdev");
+      if (ptype == registry->plugins.end()) {
+	return -ENOENT;
+      }
+
+      for (auto& i : ptype->second) {
+
+	ldout(cct, 1) << __func__ << " Trying to detect block device " << logdevname 
+		      << " using plugin " << i.first << dendl;
+	auto ebdplugin=dynamic_cast<ExtBlkDevPlugin*>(i.second);
+	if(ebdplugin==NULL){
```

Comment:
We prefer nullptr over NULL in c++ code. Unless it is a copy-paste from C logic, like fopen that has a documentation that clearly states "NULL" is returned.

Reply:
Fixed.